cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,327 @@
+                                                      clusterModel);
+
+    BrokerSetResolver brokerSetResolver = _balancingConstraint.brokerSetResolver();
+    _brokerSetResolutionHelper = new BrokerSetResolutionHelper(clusterModel, brokerSetResolver);
+    _brokersByBrokerSet = _brokerSetResolutionHelper.brokersByBrokerSetId();
+    _replicaToBrokerSetMappingPolicy = _balancingConstraint.replicaToBrokerSetMappingPolicy();
+  }
+
+  /**
+   * Update goal state.
+   * Sanity check: After completion of balancing / self-healing, confirm that replicas of each topic reside fully in a
+   * broker set.
+   *
+   * @param clusterModel        The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    // Sanity check: No self-healing eligible replica should remain at a dead broker/disk.
+    GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+    // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+    GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+    // Sanity check to confirm that the final distribution is broker set aware.
+    ensureBrokerSetAware(clusterModel, optimizationOptions);
+    if (_provisionResponse.status() != ProvisionStatus.OVER_PROVISIONED) {
+      _provisionResponse = new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
+    }
+    finish();
+  }
+
+  private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    // Sanity check to confirm that the final distribution is brokerSet aware.
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+
+    for (Map.Entry<String, List<Partition>> partitionsByTopic : clusterModel.getPartitionsByTopic().entrySet()) {
+      if (!excludedTopics.contains(partitionsByTopic.getKey())) {
+        List<Partition> partitions = partitionsByTopic.getValue();
+        Set<Broker> allBrokersForTopic = partitions.stream()
+                                                   .map(partition -> partition.partitionBrokers())
+                                                   .flatMap(brokers -> brokers.stream())
+                                                   .collect(Collectors.toSet());
+        // Checks if a topic's brokers do not all live in a single brokerSet
+        if (_brokersByBrokerSet.values().stream().noneMatch(brokerSetBrokers -> brokerSetBrokers.containsAll(allBrokersForTopic))) {
+          throw new OptimizationFailureException(
+              String.format("[%s] Topic %s is not brokerSet-aware. brokers (%s).", name(), partitionsByTopic.getKey(), allBrokersForTopic));
```

Comment:
Should we create a followup issue to add a BrokerSet-aware `ProvisionRecommendation` in case we need to add more brokers to a certain broker set?

Currently this goal does not provide a specific provision recommendation, which causes failure to automate adding new brokers to the right broker sets in case of optimization failures.

Reply:
Created follow-up issue - https://github.com/linkedin/cruise-control/issues/1842

Reply:
Created https://github.com/linkedin/cruise-control/issues/1842